File: src/main/java/pcb/uwu/core/UnitCounter.java
package pcb.uwu.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class UnitCounter {
	private final Map<Unit, Integer> major, minor;

	// region constructors

	public UnitCounter() {
		major = new HashMap<>();
		minor = new HashMap<>();
	}

	public UnitCounter(UnitCounter source) {
		major = new HashMap<>(source.major);
		minor = new HashMap<>(source.minor);
	}

	private UnitCounter(Map<Unit, Integer> major, Map<Unit, Integer> minor) {
		this.major = major;
		this.minor = minor;
	}

	// endregion

	public UnitCounter major(UnitCounter unitCounter) {
		UnitCounter result = new UnitCounter(this);

		unitCounter.major.forEach(result::majorMaps);
		unitCounter.minor.forEach(result::majorMaps);

		return result;
	}

	public UnitCounter major(Unit unit) {
		return major(unit, 1);
	}

	public UnitCounter major(Unit unit, int counts) {
		UnitCounter result = new UnitCounter(this);

		result.majorMaps(unit, result.remove(unit) + counts);

		return result;
	}

	public UnitCounter minor(UnitCounter unitCounter) {
		UnitCounter result = new UnitCounter(this);

		unitCounter.major.forEach(result::minorMaps);
		unitCounter.minor.forEach(result::minorMaps);

		return result;
	}

	public UnitCounter minor(Unit unit) {
		return minor(unit, 1);
	}

	public UnitCounter minor(Unit unit, int counts) {
		UnitCounter result = new UnitCounter(this);

		result.majorMaps(unit, result.remove(unit) - counts);

		return result;
	}

	public int get(Unit unit) {
		if (major.containsKey(unit)) {
			return major.get(unit);
		}
		if (minor.containsKey(unit)) {
			return minor.get(unit);
		}

		return 0;
	}

	public UnitCounter inverse() {
		return new UnitCounter(minor, major);
	}

	public List<Unit> getNegativeKeys() {
		return new ArrayList<>(minor.keySet());
	}

	public List<Unit> getPositiveKeys() {
		return new ArrayList<>(major.keySet());
	}

	// region private methods

	private void minorMaps(Unit unit, int counts) {
		majorMaps(unit, -counts);
	}

	private void majorMaps(Unit unit, int counts) {
		if (counts < 0) {
			minor.put(unit, counts);
		} else if (counts > 0) {
			major.put(unit, counts);
		} else {
			minor.remove(unit);
			major.remove(unit);
		}
	}

	private int remove(Unit key) {
		if (major.containsKey(key)) {
			return major.remove(key);
		}
		if (minor.containsKey(key)) {
			return minor.remove(key);
		}

		return 0;
	}

	public boolean isEmpty() {
		return major.isEmpty() && minor.isEmpty();
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/core/UnitCounter.kt
package pcb.uwu.core

import pcb.uwu.core.UnitCount.Companion.EMPTY_BASE_UNIT_COUNT
import pcb.uwu.exception.OffendingUnitException
import kotlin.math.abs
import kotlin.reflect.KClass
import kotlin.reflect.full.isSuperclassOf

class UnitCounter
{
    private val counts: MutableMap<KClass<out BaseUnit>, UnitCount>

    // region constructors

    constructor() : this(HashMap())

    constructor(unit: BaseUnit) : this()
    {
        addPower(unit.baseUnitType, unit, 1)
    }

    constructor(source: UnitCounter) : this(source.counts)

    private constructor(counts: Map<KClass<out BaseUnit>, UnitCount>)
    {
        this.counts = HashMap(counts)
    }

    // endregion

    // region public methods

    fun invert(): UnitCounter
    {
        val result: MutableMap<KClass<out BaseUnit>, UnitCount> = HashMap()

        counts.forEach { (clazz, unitCount) ->
            result[clazz] = UnitCount(unitCount.unit, -unitCount.count)
        }

        return UnitCounter(result)
    }

    fun major(unit: CompositeUnit) =
        UnitCounter(this)
            .apply {
                unit.unitCounter.counts.forEach { (clazz, unitCount) ->
                    addMajor(clazz, unitCount)
                }
            }

    fun minor(unit: CompositeUnit) =
        UnitCounter(this)
            .apply {
                unit.unitCounter.counts.forEach { (clazz, unitCount) ->
                    addMinor(clazz, unitCount)
                }
            }

    fun major(unit: BaseUnit) =
        UnitCounter(this)
            .apply {
                addPower(unit.baseUnitType, unit, 1)
            }

    fun minor(unit: BaseUnit) =
        UnitCounter(this)
            .apply {
                addPower(unit.baseUnitType, unit, -1)
            }

    operator fun get(unit: BaseUnit) =
        counts[unit.baseUnitType] ?: EMPTY_BASE_UNIT_COUNT

    val baseUnits: Collection<UnitCount>
        get() = counts.values

    fun asString(majorString: (Unit) -> String,
                 minorString: (Unit) -> String)
            : String
    {
        val result = StringBuilder()
        val baseUnits = counts.values.asSequence()
            .map { it.unit }
            .sorted()
            .toList()

        var first = true
        for (unit in baseUnits)
        {
            val power = buildPower(get(unit).count)
            if (first)
            {
                result.append(majorString(unit)).append(power)
            }
            else
            {
                result.append('⋅').append(minorString(unit)).append(power)
            }
            first = false
        }
        return result.toString()
    }

    // region private methods

    private fun buildPower(power: Int): String
    {
        var mutablePower = power
        val result = StringBuilder()
        val negative = mutablePower < 0

        // Omit neutral power of 1
        if (mutablePower == 1)
        {
            return result.toString()
        }
        if (mutablePower == 0)
        {
            return POWERS[0].toString()
        }
        mutablePower = abs(mutablePower)
        while (mutablePower > 0)
        {
            result.insert(0, POWERS[mutablePower % 10])
            mutablePower /= 10
        }
        return if (negative)
        {
            "$NEGATIVE$result"
        }
        else result.toString()
    }

    private fun addMajor(kClass: KClass<out BaseUnit>, unitCount: UnitCount) =
        addPower(kClass, unitCount.unit, unitCount.count)

    private fun addMinor(kClass: KClass<out BaseUnit>, unitCount: UnitCount) =
        addPower(kClass, unitCount.unit, -unitCount.count)

    private fun addPower(kClazz: KClass<out BaseUnit>, unit: BaseUnit, count: Int)
    {
        if (count == 0) return
        val unitCount = counts[kClazz]
        val result = if (unitCount == null)
        {
            count
        }
        else
        {
            if (unitCount.unit != unit)
            {
                throw OffendingUnitException("${unitCount.unit.singularName} and ${unit.singularName} are both ${unit.baseUnitType.simpleName}")
            }
            unitCount.count + count
        }
        if (result == 0)
        {
            counts.remove(kClazz)
        }
        else
        {
            counts[kClazz] = UnitCount(unit, result)
        }
    }

    fun <U : BaseUnit> findUnit(unitClass: KClass<U>) =
        getUnit(unitClass, baseUnits)

    @Suppress("UNCHECKED_CAST")
    private fun <U : BaseUnit> getUnit(unitClass: KClass<U>, units: Collection<UnitCount>): U?
    {
        for (unitCount in units)
        {
            val unit = unitCount.unit
            if (unitClass.isSuperclassOf(unit::class)) return unit as U
        }
        return null
    }

    val isEmpty: Boolean
        get() = counts.isEmpty()

    // endregion

    // region Object

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun equals(that: Any?): Boolean
    {
        if (that is UnitCounter)
        {
            return this.counts == that.counts
        }
        return false
    }

    override fun hashCode() =
        counts.hashCode()

    // endregion

    companion object
    {
        private const val NEGATIVE = '⁻'
        private val POWERS = charArrayOf('⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹')
    }
}
